File: src/V1/klt/analysis_runs/average_gprof.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <regex>
#include <map>
#include <vector>
#include <iomanip>
#include <algorithm>
#include <filesystem>

using namespace std;
namespace fs = std::filesystem;

struct FunctionData {
    double pct_sum = 0.0;
    double self_sum = 0.0;
    double total_sum = 0.0;
    int count = 0;
};

string normalizeName(string name) {
    // Remove leading underscores
    name = regex_replace(name, regex("^_+"), "");
    return name;
}

map<string, tuple<double, double, double>> parseGprofFile(const string &filename) {
    map<string, tuple<double, double, double>> data;
    ifstream file(filename);
    if (!file.is_open()) {
        cerr << " Error opening file: " << filename << endl;
        return data;
    }

    string line;
    // Example line:
    //  38.89      0.07     0.07       63     1.11     1.11  _convolveImageHoriz
    regex pattern(R"(\[\d+\]\s+([\d\.]+)\s+([\d\.]+)\s+([\d\.]+)(?:\s+\S+){0,3}\s+([A-Za-z0-9_:+~]+)\s+\[\d+\])");

    smatch match;

    while (getline(file, line)) {
        if (regex_search(line, match, pattern)) {
            double pct_time = stod(match[1]);
            double self_sec = stod(match[2]);
            double total_sec = stod(match[3]);
            string func = normalizeName(match[4]);
            data[func] = make_tuple(pct_time, self_sec, total_sec);
        }
    }
    return data;
}

map<string, tuple<double, double, double>> averageGprof(const vector<string> &files) {
    map<string, FunctionData> sums;

    for (const auto &file : files) {
        auto parsed = parseGprofFile(file);
        for (const auto &[func, vals] : parsed) {
            auto [pct, self, total] = vals;
            sums[func].pct_sum += pct;
            sums[func].self_sum += self;
            sums[func].total_sum += total;
            sums[func].count++;
        }
    }

    map<string, tuple<double, double, double>> averages;
    for (const auto &[func, s] : sums) {
        averages[func] = make_tuple(
            s.pct_sum / s.count,
            s.self_sum / s.count,
            s.total_sum / s.count
        );
    }
    return averages;
}

int main(int argc, char *argv[]) {
    if (argc < 4) {
        cerr << "Usage: ./average_gprof run1.txt run2.txt run3.txt" << endl;
        return 1;
    }

    vector<string> files;
    for (int i = 1; i < argc; i++) files.push_back(argv[i]);

    auto results = averageGprof(files);

    string output_file = fs::path(files[0]).parent_path().string() + "averaged_results.txt";
    ofstream out(output_file);
    if (!out.is_open()) {
        cerr << "Cannot write to " << output_file << endl;
        return 1;
    }

    out << left << setw(35) << "Function"
        << setw(15) << "Avg % Time"
        << setw(20) << "Avg Self(s)"
        << setw(20) << "Avg Total(s)" << "\n";
    out << string(90, '=') << "\n";

    vector<pair<string, tuple<double, double, double>>> sorted(results.begin(), results.end());
    sort(sorted.begin(), sorted.end(), [](auto &a, auto &b) {
        return get<0>(a.second) > get<0>(b.second); // Sort by % time
    });

    cout << "\n Top Hotspot Functions (by % time):\n";
    int limit = min(10, (int)sorted.size());
    for (int i = 0; i < limit; i++) {
        cout << setw(25) << left << sorted[i].first
             << " %time_avg=" << fixed << setprecision(2) << get<0>(sorted[i].second)
             << "% self_avg=" << get<1>(sorted[i].second)
             << "s total_avg=" << get<2>(sorted[i].second) << "s\n";
    }

    for (auto &[func, vals] : sorted) {
        out << left << setw(35) << func
            << setw(15) << fixed << setprecision(2) << get<0>(vals)
            << setw(20) << get<1>(vals)
            << setw(20) << get<2>(vals) << "\n";
    }

    cout << "\n Results saved to: " << output_file << endl;
    return 0;
}

File: src/V1/klt/analysis_runs/average_gprof.py
import sys
import os
import subprocess
from collections import defaultdict
import re

def demangle(name):
    try:
        out = subprocess.run(['c++filt', name], capture_output=True, text=True)
        return out.stdout.strip()
    except Exception:
        return name

def normalize_func_name(name):
    name = name.strip()
    if name.startswith("_"):
        name = name.lstrip("_")
    if name.startswith("std::"):
        name = name.replace("std::", "")
    return name

def parse_gprof_file(filename):
    """Parse gprof file robustly by splitting on whitespace."""
    data = {}

    with open(filename, "r", encoding="utf-8", errors="ignore") as f:
        for line in f:
            if not line.strip() or line.startswith("Each sample") or "name" in line:
                continue  # skip headers and empty lines

            parts = line.split()
            if len(parts) < 7:
                continue  # not a valid data line

            try:
                percent = float(parts[0])
                cumulative = float(parts[1])
                self_sec = float(parts[2])
                # function name is always last column
                func_name = parts[-1]
                func_name = demangle(normalize_func_name(func_name))
                data[func_name] = {
                    "percent": percent,
                    "self_sec": self_sec,
                    "cum_sec": cumulative
                }
            except ValueError:
                print(f"Warning: Could not parse line in {filename}: {line.strip()}")
                #continue  # skip lines that don’t match numeric structure
    return data

def average_gprof(files):
    sums = defaultdict(lambda: {"percent": 0.0, "self_sec": 0.0, "cum_sec": 0.0})
    counts = defaultdict(int)

    for file in files:
        file_data = parse_gprof_file(file)
        for func, vals in file_data.items():
            sums[func]["percent"] += vals["percent"]
            sums[func]["self_sec"] += vals["self_sec"]
            sums[func]["cum_sec"] += vals["cum_sec"]
            counts[func] += 1

    averages = {}
    for func in sums:
        n = counts[func]
        avg_percent = sums[func]["percent"] / n
        avg_self = sums[func]["self_sec"] / n
        avg_cum = sums[func]["cum_sec"] / n
        hotspot_score = avg_percent + (avg_self * 100)
        averages[func] = (avg_percent, avg_self, avg_cum, hotspot_score)
    return averages

def save_results(results, output_file):
    with open(output_file, "w") as f:
        f.write("Function\tAvg %Time\tAvg Self (s)\tAvg Cumulative (s)\tHotspot Score\n")
        f.write("=" * 90 + "\n")
        for func, (pct, self_s, cum_s, score) in sorted(results.items(), key=lambda x: x[1][3], reverse=True):
            f.write(f"{func}\t{pct:.2f}\t{self_s:.4f}\t{cum_s:.4f}\t{score:.2f}\n")
    print(f"\n✅ Averaged results saved to: {output_file}")

def main():
    if len(sys.argv) < 4:
        print("Usage: python3 average_gprof_hotspots.py run1.txt run2.txt run3.txt")
        sys.exit(1)

    input_files = sys.argv[1:]
    results = average_gprof(input_files)
    
    output_file = os.path.join(os.path.dirname(input_files[0]), "averaged_results.txt")
    save_results(results, output_file)

    print("\n Top Hotspots (by Hotspot Score):")
    for func, (pct, self_s, cum_s, score) in sorted(results.items(), key=lambda x: x[1][3], reverse=True)[:10]:
        print(f"{func:<35} score={score:.2f}  avg_self={self_s:.4f}s  avg_pct={pct:.2f}%")

if __name__ == "__main__":
    main()
